compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1931,7 +1932,10 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         || symInfo.isInstanceOf[MethodType]
             && symInfo.signature.consistentParams(info2.signature)
 
-      def tp1IsSingleton: Boolean = tp1.isInstanceOf[SingletonType]
+      def allowGadt = mbr match
+        case _ if tp1.isInstanceOf[SingletonType]                                   => false
+        case d: UniqueRefDenotation if d.prefix == NoPrefix && d.symbol != NoSymbol => false
```

Comment:
@odersky this is what I found works.  Can you think of another way to identify when we have a member that required widening `tp1` to a non-class info and thus we can't safely infer gadt constraints?

Reply:
I can't make sense of this condition and  don't understand the logic behind this. This would at least need a fat comment explaining the issue and the attempted solution.